common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,312 @@
+
+                        state = std::make_tuple(code, keyDown, dt);
+                        break;
+                    }
+                }
+
+                return state;
+            };
+
+            const auto&& [physicalKey, isDown, dtAction] = updateVirtualState();
+
+            if (physicalKey != ui::E_KEY_CODE::EKC_NONE)
+                if (isDown)
+                    output.emplace_back(CVirtualEvent{ virtualKey, static_cast<float64_t>(dtAction) });
+        }
+
+        return output;
+    }
+
+    /*
+    // [OPTIONAL]: process mouse to generate virtual manipulation events
+    // note that:
+    // - all manipulations *may* be done with keyboard keys (if you have a touchpad or sth an ui:: event could be a code!)
+    // - it doesn't make the manipulation itself!
+    */
+    std::vector<CVirtualEvent> processMouse(std::span<const ui::SMouseEvent> events) const
+    {
+        double dPitch = {}, dYaw = {};
+
+        for (const auto& ev : events)
+            if (ev.type == nbl::ui::SMouseEvent::EET_MOVEMENT)
+            {
+                dYaw += ev.movementEvent.relativeMovementX;
+                dPitch += ev.movementEvent.relativeMovementY;
+            }
+
+        std::vector<CVirtualEvent> output;
+
+        if (dPitch)
+        {
+            auto& pitch = output.emplace_back();
+            pitch.type = dPitch > 0.f ? TiltUp : TiltDown;
+            pitch.value = std::abs(dPitch);
+        }
+
+        if (dYaw)
+        {
+            auto& yaw = output.emplace_back();
+            yaw.type = dYaw > 0.f ? PanRight : PanLeft;
+            yaw.value = std::abs(dYaw);
+        }
+
+        return output;
+    }
+
+    inline void setMoveSpeed(const float moveSpeed) { m_moveSpeed = moveSpeed; }
+    inline void setRotateSpeed(const float rotateSpeed) { m_rotateSpeed = rotateSpeed; }
+
+    inline const float getMoveSpeed() const { return m_moveSpeed; }
+    inline const float getRotateSpeed() const { return m_rotateSpeed; }
+
+protected:
+    // controller can override default set of event map
+    virtual void initKeysToEvent()
+    {
+        m_keysToEvent[MoveForward] = ui::E_KEY_CODE::EKC_W ;
+        m_keysToEvent[MoveBackward] =  ui::E_KEY_CODE::EKC_S ;
+        m_keysToEvent[MoveLeft] =  ui::E_KEY_CODE::EKC_A ;
+        m_keysToEvent[MoveRight] =  ui::E_KEY_CODE::EKC_D ;
+        m_keysToEvent[MoveUp] =  ui::E_KEY_CODE::EKC_SPACE ;
+        m_keysToEvent[MoveDown] =  ui::E_KEY_CODE::EKC_LEFT_SHIFT ;
+        m_keysToEvent[TiltUp] =  ui::E_KEY_CODE::EKC_NONE ;
+        m_keysToEvent[TiltDown] =  ui::E_KEY_CODE::EKC_NONE ;
+        m_keysToEvent[PanLeft] =  ui::E_KEY_CODE::EKC_NONE ;
+        m_keysToEvent[PanRight] =  ui::E_KEY_CODE::EKC_NONE ;
+        m_keysToEvent[RollLeft] =  ui::E_KEY_CODE::EKC_NONE ;
+        m_keysToEvent[RollRight] =  ui::E_KEY_CODE::EKC_NONE ;
+        m_keysToEvent[Reset] =  ui::E_KEY_CODE::EKC_R ;
+    }
+
+    core::smart_refctd_ptr<CGimbal> m_gimbal;
```

Comment:
don't make the gimbal live in the main camera interface, some cameras may have multiple gimbals

Reply:
resolved